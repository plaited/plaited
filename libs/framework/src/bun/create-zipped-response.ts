export const jsMimeTypes = ['text/javascript;charset=utf-8', 'application/json;charset=utf-8'] as const

export const isJsMimeType = (type: string): type is (typeof jsMimeTypes)[number] =>
  jsMimeTypes.includes(type as (typeof jsMimeTypes)[number])

export const createZippedResponse = (str: string, type: (typeof jsMimeTypes)[number]) => {
  const compressed = Bun.gzipSync(Buffer.from(str))
  return new Response(compressed, {
    headers: {
      'content-type': type,
      'content-encoding': 'gzip',
    },
  })
}
