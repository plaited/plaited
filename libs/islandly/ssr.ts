import { Template } from './create-template.ts'
export const ssr = (...templates: Template[]) => {
  return '<!DOCTYPE html> ' + templates.map((tpl) => tpl.content).join('')
}
