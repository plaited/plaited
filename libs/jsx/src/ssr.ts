import { TemplateObject } from '@plaited/component-types'
import { escape, isTypeOf } from '../../../libs-cross-framework/utils/dist/index.js'
import { validPrimitiveChildren } from './constants.js'
if (typeof global.HTMLElement === 'undefined') {
  // @ts-ignore node env
  global.HTMLElement = class HTMLElement {}
}

export const ssr = (...templates: TemplateObject[]) => {
  const arr = []
  const stylesheets = new Set<string>()
  const length = templates.length
  for (let i = 0; i < length; i++) {
    const child = templates[i]
    if (isTypeOf<Record<string, unknown>>(child, 'object') && child.$ === 'ðŸ¦„') {
      arr.push(...child.server)
      for (const sheet of child.stylesheets) {
        stylesheets.add(sheet)
      }
      continue
    }
    if (!validPrimitiveChildren.has(typeof child)) continue
    const safeChild = escape(`${child}`)
    arr.push(safeChild)
  }
  const style = stylesheets.size ? `<style>${[...stylesheets].join('')}</style>` : ''
  const str = arr.join('')
  const headIndex = str.indexOf('</head>')
  const bodyRegex = /<body\b[^>]*>/i
  const bodyMatch = bodyRegex.exec(str)
  const bodyIndex = bodyMatch ? bodyMatch.index + bodyMatch[0].length : 0

  const index = headIndex !== -1 ? headIndex : bodyIndex

  return str.slice(0, index) + style + str.slice(index)
}
