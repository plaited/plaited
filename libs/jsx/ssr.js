export * from './dist/ssr.js'
