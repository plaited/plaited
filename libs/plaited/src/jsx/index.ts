export * from './create-template.js'
