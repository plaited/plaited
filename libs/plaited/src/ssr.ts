export { ssr } from './jsx/ssr.js'
