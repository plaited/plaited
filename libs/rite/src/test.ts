/// <reference lib="dom" />
/// <reference lib="dom.iterable" />

import { TestResultError, TestResult, BrowserSessionResult } from '@web/test-runner-core/browser/session.js'
import { assert, Assertion, AssertionError } from './assert.js'

type TestCallback = (t: Assertion) => Promise<void> | void

interface Test {
  (name: string, cb: TestCallback): void
  skip: (name: string, cb: TestCallback) => void
}

type TestCase = () => Promise<boolean> | boolean

class TestRunner {
  #completedTests = 0
  #failedTests = 0
  #errors: TestResultError[] = []
  #cases: TestCase[] = []
  #timeout = 5_000
  #tests: TestResult[] = []
  updateTimeout(time: number) {
    this.#timeout = time
  }
  get timeout() {
    return this.#timeout
  }
  addCase(test: TestCase) {
    this.#cases.push(test)
  }
  addResult(result: TestResult) {
    this.#tests.push(result)
  }
  addError(obj: TestResultError) {
    this.#errors.push(obj)
  }
  async run(name: string) {
    for (const testCase of this.#cases) {
      try {
        const passed = await testCase()
        this.#completedTests++
        if (!passed) {
          this.#failedTests++
        }
      } catch (error) {
        console.error(error)
      }
    }

    if (this.#completedTests === this.#cases.length) {
      const results: BrowserSessionResult = {
        passed: !this.#failedTests,
        testResults: {
          name,
          suites: [],
          tests: this.#tests,
        },
        errors: this.#errors,
      }
      return results
    } else {
      throw new AssertionError(`Not all test completed`)
    }
  }
}

declare global {
  interface Window {
    __rite_test_runner?: TestRunner
  }
}

window.__rite_test_runner = new TestRunner()

const throwTimeoutError = async () => {
  const timeout = window.__rite_test_runner.timeout
  await assert.wait(timeout)
  throw new Error(`test takes longer than ${timeout}ms to complete`)
}

export const test: Test = (name, callback) => {
  window.__rite_test_runner.addCase(async () => {
    const startTime = performance.now()
    const result: TestResult = {
      name,
      passed: true,
      skipped: false,
    }
    try {
      await Promise.race([callback(assert), throwTimeoutError()])
      const msg = `✓ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
    } catch (error) {
      const msg = `✗ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
      result.passed = false
      if (error instanceof AssertionError) {
        const obj: Pick<TestResultError, 'actual' | 'expected' | 'message'> = JSON.parse(error.message)
        result.error = { name, stack: `${error.stack}`, ...obj }
        window.__rite_test_runner.addError({ name, ...obj })
      } else {
        throw new Error(error)
      }
    }
    result.duration = performance.now() - startTime
    window.__rite_test_runner.addResult(result)
    return result.passed
  })
}
test.skip = (name: string, _: TestCallback) => {
  window.__rite_test_runner.addCase(() => {
    const result = {
      name,
      passed: true,
      skipped: true,
    }
    window.__rite_test_runner.addResult(result)
    return true
  })
}
