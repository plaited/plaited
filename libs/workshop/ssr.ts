import { element } from '../island/mod.ts'
import { SSRFunc } from './types.ts'

export const ssr: SSRFunc = (tag, template) => element({ tag, template })
