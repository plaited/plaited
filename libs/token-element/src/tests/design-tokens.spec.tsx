import { test, expect } from 'bun:test'
import { getTokenElement } from '../index.js'
import { css } from '@plaited/jsx'
import { ssr } from '@plaited/jsx/ssr'
import beautify from 'beautify'

test('getTokenElement', () => {
  const [_, stylesheet] = css`
    :host {
      --blue: blue;
    }
  `
  const DesignTokens = getTokenElement(stylesheet)
  const Template = DesignTokens.template

  expect(beautify(ssr(<Template />), { format: 'html' })).toMatchSnapshot('no children')
  expect(
    beautify(
      ssr(
        <Template>
          <div>hello world!</div>
        </Template>,
      ),
      { format: 'html' },
    ),
  ).toMatchSnapshot('with children')
})
