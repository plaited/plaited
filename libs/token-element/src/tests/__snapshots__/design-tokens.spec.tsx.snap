// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getTokenElement: no children 1`] = `
"<design-tokens>
    <template shadowrootmode="open" shadowrootdelegatesfocus="true">
        <style>
            :host { display: contents; }:host { --blue: blue; }
        </style>
        <slot></slot>
    </template>
</design-tokens>"
`;

exports[`getTokenElement: with children 1`] = `
"<design-tokens>
    <template shadowrootmode="open" shadowrootdelegatesfocus="true">
        <style>
            :host { display: contents; }:host { --blue: blue; }
        </style>
        <slot></slot>
    </template>
    <div>hello world!</div>
</design-tokens>"
`;
